### PR TITLE
Support inheritance

### PR DESCRIPTION
**Pull request by VisualSJ**
Support inheritance

1. profile.register - adds a third parameter
2. The schema can only be set on the root profile
3. profile.get - Look up the data in turn

@jwu

### CODE REVIEW

File: lib/main.js [hunk 1]
```diff
@@ -111,14 +113,22 @@ profile.load = function ( url ) {
...
  *
  * Register profile type with the path you provide.
  * {{#crossLink "profile.load"}}{{/crossLink}}
  */
-profile.register = function ( type, path ) {
+profile.register = function ( type, path, inherit ) {
   _type2path[type] = path;
 };
 
+profile.inherit = function ( type, parent ) {
```
Comment (jwu) [inline lib/main.js]: comments please :D

File: lib/main.js [hunk 2]
```diff
@@ -146,24 +156,57 @@ profile.registerSchema = function ( url, json ) {
...
+
+  if ( url.indexOf('profile://') !== 0 ) {
+    console.error(`Failed to load profile ${url}: invalid protocol.`);
+    return;
+  }
+
+  let type = url.match(/^profile:\/\/([^\/]+)/)[1];
+
+  if (_type2parent[type]) {
+    console.log(`Warnning to register schema ${url}: The schema only takes effect on the root type`);
```
Comment (jwu) [inline lib/main.js]: console.warn(`Failed to register schema ${url}: only root profile can register schema`);

File: lib/main.js [hunk 3]
```diff
@@ -146,24 +156,57 @@ profile.registerSchema = function ( url, json ) {
...
+  }
   _url2schema[url] = schema.json2schema(json);
 };
 
 // ==========================
 // Internal
 // ==========================
 
+let _searchRootType = function (type) {
```
Comment (jwu) [inline lib/main.js]: _getRootType

File: lib/main.js [hunk 4]
```diff
@@ -146,24 +156,57 @@ profile.registerSchema = function ( url, json ) {
...
+
+    let match = url.match(/^profile\:\/\/([^\/]+)\/(.*)$/);
+    this._type = match[1];
+    this._file = match[2];
   }
 
   get (key) {
-    let schema = _url2schema[this._url];
+    let root = _searchRootType(this._type);
+    let schema = _url2schema[`profile://${root}/${this._file}`];
```
Comment (jwu) [inline lib/main.js]: how about `_getSchema(this.type)`

File: lib/main.js [hunk 5]
```diff
@@ -146,24 +156,57 @@ profile.registerSchema = function ( url, json ) {
...
     let target = this._data[key];
 
+    // Find the data on the inheritance chain
+    let type = this._type;
+    while (target === undefined && _type2parent[type]) {
+      let inherit = _type2parent[type];
```
Comment (jwu) [inline lib/main.js]: inherit => parent

**Comment by VisualSJ** [comment]
@jwu modified

**Pull request merged**